AWS Serverless Analytics -- blog post (#2305)
pulumi/docs: content/blog/aws-serverless-analytics/index.md

---
title: "AWS Serverless Analytics"
authors: ["evan-boyle"]
tags: ["AWS-Serverless", "Analytics", "Architecture-as-Code", "Data-Warehouse"]
date: "2020-01-28"
meta_desc: "Building a serverless data warehouse on AWS using architecture as code."
meta_image: "ServerlessArchitecture.png"
---

Whether it’s an IoT installation, a website, or a mobile app, modern software systems generate a trove of usage and performance data. While it can be daunting to collect and manage, surfacing data empowers the business to make informed product investments. In this article, we’ll explore the following:

1. An overview of the traditional Redshift analytics stack on AWS, the use cases it excels at, and where it falls apart.
2. An alternative architecture utilizing serverless and streaming.
3. How to implement this architecture as code in a reusable library with Pulumi.
4. How to automate the development loop when writing Pulumi libraries.

> If you’d like to follow along, you can clone and run the reference implementation (TODO add link once examples PR is merged). If you’re new to Pulumi, you can follow [this guide to get started](/docs/get-started/).

Author: Done

3. __Ad-hoc workloads interfere with production__: Business users write sophisticated queries that accidentally bring the cluster to a standstill. The team puts a complicated [WLM policy](https://docs.aws.amazon.com/redshift/latest/dg/tutorial-wlm-routing-queries-to-queues.html) in place to try and isolate ad-hoc from production, but this is more of an art than a science.

## Designing the V2 Analytics Stack

The team understands how to keep the analytics stack limping along at this point, but the cost in both hardware and developer time is too high. They decide to pursue a redesign with the following goals in mind:

1. Little to no operator intervention required. Developer time is their most valuable asset, and going forward, they want to focus engineering on delivering new features to advertisers, and new experiences to customers.
2. Scale the analytics stack to 1B events per day and beyond as the program grows in terms of advertisers, and geography.
3. Enable workload isolation between production services and ad-hoc business intelligence users.

![Serverless Data Warehouse](./ServerlessArchitecture.png)

> JSON events are queued in Kinesis, consumed by a Firehose that validates their schema against the definition in Glue, then converted into [parquet](https://parquet.apache.org/), and finally stored in S3 in hourly folders based on arrival time. Structured data in S3 and a Glue metastore are flexible primitives that end-users and production systems can build on. Both production systems and ad-hoc users can bring their own compute or take advantage of serverless solutions like Athena (the AWS serverless version of [Presto](https://prestodb.io/)) to query over the data with isolation. The operator overhead is low because it relies almost entirely on serverless offerings. This architecture brings some additional benefits:

Author: 👍

While our new serverless streaming architecture brings a slew of benefits that enable ease of operation and the next level of scale, there are some drawbacks. For one, there is substantially more infrastructure to configure than the single Redshift cluster. How do we implement this in a repeatable fashion?

## Architecture as Code with Pulumi

We want repeatable infrastructure that we can reliably develop, test, and deploy. Creating a single Redshift instance through the console was manageable, but our serverless architecture utilizes multiple instances of five AWS services. Serverless solutions frequently require many more granular resources, and thus the complexity of the infrastructure as code increases, and we need more robust tools to manage this complexity - just like we manage software complexity.

> Pulumi takes care of the complexity by bringing software development best practices like abstraction, encapsulation, and modularity to infrastructure. You can use design patterns in your favorite programming language to provision modern cloud infrastructure. [Component Resources](https://www.pulumi.com/blog/creating-and-reusing-cloud-components-using-package-managers/) are our tools for encapsulating low-level cloud infrastructure into Architecture as Code that is shareable across teams.

Reviewer: ```suggestion
Pulumi takes care of the complexity by bringing software development best practices like abstraction, encapsulation, and modularity to infrastructure. You can use design patterns in your favorite programming language to provision modern cloud infrastructure. [Component Resources]({{< rel "/blog/creating-and-reusing-cloud-components-using-package-managers/index.md" >}}) are our tools for encapsulating low-level cloud infrastructure into Architecture as Code that is shareable across teams.
```
Use Hugo shortcode for links to internal

Author: 👍

const tableArgs: StreamingInputTableArgs = {
    columns,
    inputStreamShardCount: 1,
    region: "us-west-2",
};

// create two tables with kinesis input streams,
// writing data into hourly partitions in S3.
const dataWarehouse = new ServerlessDataWarehouse("analytics_dw", { isDev })
    .withStreamingInputTable(impressionsTableName, genericTableArgs)
    .withStreamingInputTable(clicksTableName, genericTableArgs);

// Make input streams available to external programs as stack outputs
export const impressionInputStream = impressionsInputStream.name;
export const clickInputStream = clicksInputStream.name;
```

In a dozen lines of code, we’ve provisioned two of our desired tables, “clicks” and “impressions”, using our serverless streaming input architecture. It takes care of everything, including implicitly creating an arrival time [partition scheme](https://docs.aws.amazon.com/athena/latest/ug/partitions.html) with the key “inserted_at”. We export our Kinesis input streams as [stack outputs](/docs/intro/concepts/stack/#outputs) that can be referenced in other projects, such as the instrumentation within our ad server or consumer-facing web app.

> While on the surface this Pulumi component is described imperatively, it produces a declarative output in the form of a [state file](/docs/intro/concepts/state/) that can be managed locally, in an object store like S3, or by the Pulumi Service backend. Running a ‘pulumi up’ shows that we’ve created 45 AWS resources, and lists our stack outputs to the console.

Reviewer: ```suggestion
While on the surface this Pulumi component is described imperatively, it produces a declarative output in the form of a [state file]({{<  rel "docs/intro/concepts/statemd" >}}) that can be managed locally, in an object store like S3, or by the Pulumi Service backend. Running a ‘pulumi up’ shows that we’ve created 45 AWS resources, and lists our stack outputs to the console.
```

Hugo uses `ref` and `relref` shortcodes to render internal links

Author: 👍

```typescript
export class ServerlessDataWarehouse extends pulumi.ComponentResource {

    public dataWarehouseBucket: aws.s3.Bucket;
    public queryResultsBucket: aws.s3.Bucket;
    public database: aws.glue.CatalogDatabase;
    private tables: { [key: string]: aws.glue.CatalogTable } = {};
    private inputStreams: { [key: string]: aws.kinesis.Stream } = {};

    constructor(name: string,
        args: DataWarehouseArgs = {},
        opts: pulumi.ComponentResourceOptions = {}
    ) {
        super("serverless:data_warehouse", name, opts);

        const bucketArgs: BucketArgs | undefined = args.isDev ?
            { forceDestroy: true } : undefined;

        const dataWarehouseBucket = new aws.s3.Bucket("datawarehouse-bucket",
            bucketArgs, { parent: this });
        const queryResultsBucket = new aws.s3.Bucket("query-results-bucket",
            bucketArgs, { parent: this });

        const database = args.database || new aws.glue.CatalogDatabase(name, {
            name,
        }, { parent: this });

        this.dataWarehouseBucket = dataWarehouseBucket;
        this.queryResultsBucket = queryResultsBucket;
        this.database = database;
    }
}
```

> The `ServerlessDataWarehouse` class creates a shared S3 data bucket, Athena query results bucket, and glue database. We’ve exposed a fluent method, `.withStreamingInputTable()`, that creates the Glue table, Kinesis input stream, Firehose destination stream, and the Lambda cron that registers the hourly partitions. Many of these pieces themselves are also encapsulated into higher-order components.

Reviewer: ```suggestion
The `ServerlessDataWarehouse` class creates a shared S3 data bucket, Athena query results bucket, and Glue database. We’ve exposed a fluent method, `.withStreamingInputTable()`, that creates the Glue table, Kinesis input stream, Firehose destination stream, and the Lambda cron that registers the hourly partitions. Many of these pieces themselves are also encapsulated into higher-order components.
```

Author: 👍

    const partitionKeys: input.glue.CatalogTablePartitionKey[] = [{
        name: partitionKey,
        type: "string",
    }];

    const tableArgs: TableArgs = {
        columns,
        partitionKeys,
    };

    this.withTable(name, tableArgs);

    const streamArgs: InputStreamArgs = {
        destinationBucket: this.dataWarehouseBucket,
        shardCount: inputStreamShardCount,
        databaseName: this.database.name,
        tableName: name,
        fileFlushIntervalSeconds,
    };

    const { inputStream } = new InputStream(`inputstream-${name}`,
        streamArgs, { parent: this });
    this.inputStreams[name] = inputStream;

    const registrarArgs: PartitionRegistrarArgs = {
        database: this.database,
        partitionKey,
        region: args.region,
        dataWarehouseBucket: this.dataWarehouseBucket,
        athenaResultsBucket: this.queryResultsBucket,
        table: name,
        scheduleExpression: args.partitionScheduleExpression,
    };
    const partitionRegistrar = new HourlyPartitionRegistrar(
        `${name}-partitionregistrar`, registrarArgs, { parent: this });

    return this;
}
```

> Our withStreamingInputTable call makes use of another fluent API, `.withTable()`, that encapsulates the logic around various storage and SerDe details required for storing different formats in glue (Parquet vs. JSON) in a way that Athena can understand:

Reviewer: ```suggestion
Our `withStreamingInputTable` call makes use of another fluent API, `.withTable()`, that encapsulates the logic around various storage and SerDe details required for storing different formats in glue (Parquet vs. JSON) in a way that Athena can understand:
```

Author: 👍

    // write the event counts to our table in S3 in JSON format
    const clickCount = clickRows.records[0]["_col0"];
    const impressionsCount = impressionRows.records[0]["_col0"];
    const data = `{ "event_type": "${clicksTableName}", `+
        `"count": ${clickCount}, "time": "${partitionKey}" }\n` +
        `{ "event_type": "${impressionsTableName}", ` +
        `"count": ${impressionsCount}, "time": "${partitionKey}"}`;
    const s3Client = new S3();
    await s3Client.putObject({
        Bucket: dwBucket.get(),
        Key: `${aggregateTableName}/${partitionKey}/results.json`,
        Body: data,
    }).promise();
};

//configure resource access policies required by the function
const policyARNsToAttach: pulumi.Input<ARN>[] = [
    aws.iam.ManagedPolicies.AmazonAthenaFullAccess,
    aws.iam.ManagedPolicies.AmazonS3FullAccess,
];

const aggregateTableArgs: BatchInputTableArgs = {
    columns: aggregateTableColumns,
    jobFn: aggregationFunction,
    scheduleExpression,
    policyARNsToAttach,
    dataFormat: "JSON",
};

dataWarehouse.withBatchInputTable(aggregateTableName, aggregateTableArgs);
```

![Athena Aggregate Table Query Results](./AggregatesExample.png)

> While Lambda’s limits restrict the utility of this method, the concept is powerful. Pulumi gives you the building blocks to define your abstractions. Rather than Lambda, the execution plane could be ECS, Kubernetes, or EMR.

Reviewer: ```suggestion
While Lambda’s limits restrict the utility of this method, the concept is powerful. Pulumi gives you the building blocks to define your abstractions. Instead of Lambda, the execution plane could be ECS, Kubernetes, or EMR.
```

Author: 👍

Here we’ve seen three different ways to use Pulumi to automate and improve the inner dev loop. Simple scripting with stack outputs to get started, automation of that script using Pulumi, and full blow integration testing to bring 100% confidence in tricky changes. All three are tools in our arsenal to be applied appropriately.

## Wrapping Up

ACME re-evaluated their analytics architecture and shipped the next-generation serverless streaming solution that can take them to the next click-stop, and well through the next order of magnitude. Along the way, they learned how to reason about streaming data and examined the underlying assumptions about time and completeness of data in their existing system. ACME developed its streaming architecture using Pulumi and learned how to automate and create reproducible developer environments along the way. The outcome isn’t just an instance of the serverless streaming architecture, but a component library that implements best practices. That investment in time is bundled up and shipped in a format that the next generation of incubators at ACME can pick up and run with.

> If you’d like to get started delivering best practices in infrastructure across your organization, [get started with Pulumi free today](/docs/get-started). You can find the reference implementation of the Serverless Data Warehouse ready to clone and modify here (TODO link after examples PR merged).

Reviewer: ```suggestion
If you’d like to get started delivering best practices in infrastructure across your organization, [get started with Pulumi free today]({{< rel "/docs/get-started/_index.md" >}}). You can find the reference implementation of the Serverless Data Warehouse ready to clone and modify here (TODO link after examples PR merged).
```

add link to examples

Author: Done.
